### PR TITLE
[fix][client] Close deadletter topic producer and retry topic producer

### LINKED ISSUE #14284: when Consumer close,retryLetterProducer and deadLetterProducer will not close
**Describe the bug**
when Consumer close,retryLetterProducer and deadLetterProducer will not close, it will produce 

Topic reached max producers limit  for retry topic sometimes 

**To Reproduce**

review the code 

**Expected behavior**
when Consumer close,retryLetterProducer and deadLetterProducer will close

**Screenshots**

**Desktop (please complete the following information):**

**Additional context**

### PR DESCRIPTION
**Pull request by leizhiyuan**
Fixes #14284

*(or if this PR is one task of a github issue, please add `Master Issue: #<xyz>` to link to the master issue.)*

Master Issue: #<xyz>

### Motivation

When the consumer was closed, the producer of delay topic and dead topic were not closed, so when consumer closed, I will close deadProducer and retryProducer (if they exist, maybe this consumer does not have deadProducer and  retryProducer)

I will wait for the retryProducer and deadProducer closed . 

### Modifications

*Describe the modifications you've done.*

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API: (yes / no)
  - The schema: (yes / no / don't know)
  - The default values of configurations: (yes / no)
  - The wire protocol: (yes / no)
  - The rest endpoints: (yes / no)
  - The admin cli options: (yes / no)
  - Anything that affects deployment: (yes / no / don't know)

### Documentation

Check the box below or label this PR directly (if you have committer privilege).

Need to update docs? 

- [ ] `doc-required` 
  
  (If you need help on updating docs, create a doc issue)
  
- [x] `no-need-doc` 
  
  (Please explain why)
  
- [ ] `doc` 
  
  (If this PR contains doc changes)

- [x] `doc-not-needed`

### CODE REVIEW

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java [hunk 1]
```diff
@@ -1008,6 +1008,20 @@ public void connectionFailed(PulsarClientException exception) {
 
         closeConsumerTasks();
 
+        if (this.deadLetterProducer != null) {
+            try {
+                deadLetterProducer.get().closeAsync();
```
Comment (RobertIndie) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java:1013]: Please use asynchronous calls here and make it to be connected to `closeFuture`.

**Review by RobertIndie — COMMENTED** [review]
Hi @leizhiyuan Thanks for your PR. Could you add some tests to verify this fix?

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java [hunk 2]
```diff
@@ -1021,12 +1024,34 @@ public void connectionFailed(PulsarClientException exception) {
...
+        closeFutures.add(closeConsumerFuture);
+        return CompletableFuture.allOf(closeFutures.toArray(new CompletableFuture[0]));
+    }
 
-        return closeFuture;
+    private List<CompletableFuture<Void>> closeDeadAndRetryProducer() {
+        final List<CompletableFuture<Void>> closeFutures = new ArrayList<>();
+        if (this.deadLetterProducer != null) {
+            try {
+                deadLetterProducer.whenComplete((r, t) -> {
```
Comment (Technoboy-) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java:1039]: deadLetterProducer.closeAsync()?

Comment (leizhiyuan) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java:1039]: deadLetterProducer is future ,not producer

**Review by michaeljmarshall — COMMENTED** [review]
I tried to fix this issue in this PR https://github.com/apache/pulsar/pull/10159, but I wasn't able to complete it at that time. Please take a look at the feedback there, too. 

I'll try to review this soon, too. Thanks.

**Comment by michaeljmarshall** [comment]
@leizhiyuan - can you please edit the PR's description to provide detail for this change? Thanks.

**Comment by leizhiyuan** [comment]
> I tried to fix this issue in this PR #10159, but I wasn't able to complete it at that time. Please take a look at the feedback there, too.
> 
> I'll try to review this soon, too. Thanks.

OK, I didn't notice that pr before. I will take a look.

**Comment by BewareMyPower** [comment]
Could you continue this work?

**Comment by Jason918** [comment]
@leizhiyuan ping

**Comment by leizhiyuan** [comment]
any suggestion?

**Comment by congbobo184** [comment]
@leizhiyuan   hi, I move this PR to release/2.9.5, if you have any questions, please ping me. thanks.

**Comment by tisonkun** [comment]
> Could you continue this work?

@BewareMyPower @Jason918 I've rebased this patch and I think it's ready for review already.

I'm looking into this patch now.

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java [hunk 3]
```diff
@@ -1049,12 +1052,34 @@ public CompletableFuture<Void> closeAsync() {
                 if (ignoreException && exception != null) {
                     log.debug("Exception ignored in closing consumer", exception);
                 }
-                cleanupAtClose(closeFuture, ignoreException ? null : exception);
+                cleanupAtClose(closeConsumerFuture, ignoreException ? null : exception);
                 return null;
             });
         }
+        closeFutures.add(closeConsumerFuture);
+        return CompletableFuture.allOf(closeFutures.toArray(new CompletableFuture[0]));
```
Comment (RobertIndie) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java:1060]: Please use `FutureUtil.waitForAll`

**Comment by RobertIndie** [comment]
Is it possible to add some tests to verify it?

**Comment by tisonkun** [comment]
It seems https://github.com/apache/pulsar/pull/19858 covers this patch.

Closing...

Thanks for your time invested @leizhiyuan!

**Comment by michaeljmarshall** [comment]
As discussed on the mailing list https://lists.apache.org/thread/w4jzk27qhtosgsz7l9bmhf1t7o9mxjhp, there is no plan to release 2.9.6, so I am going to remove the release/2.9.6 label

**Pull request closed without merging**